### ISSUE TITLE
Implement a 'Wikipedia' login provider

### ISSUE THREAD

**Issue by yuvipanda** [sandstorm-io/sandstorm#525]
It has OAuth support that can be used to implement a login provider now, and might be a more 'privacy conscious'{{cn}} alternative to Google / GitHub.

**Comment by paulproteus** [comment]
As context, for people not familiar with Wikipedia culture:

{{cn}} is short-hand for "Citation needed," which is Yuvi Panda's way of
saying that the idea that Wikipedia (hosted by the Wikimedia Foundation)
login is more private than Google/GitHub possibly needs further
justification.

Also for what it's worth for Yuvi -- the way to do this would be to make
sure there is a Meteor-accounts plugin for Wikipedia login, and then write
a small bit of code to make that an option for Sandstorm installs.

**Comment by ocdtrekkie** [comment]
Well, Wikipedia accounts are, as far as I know, less explicitly tied to real identities than Google or Github. And Wikipedia, as far as I know, doesn't try to profile you on other sites, which makes it likely for that reason a step above Google.

The big one that might be blocking for @kentonv is that I've never heard of two-factor authentication for Wikipedia accounts.

**Comment by yuvipanda** [comment]
Wikipedia does not support two factor authentication, true. And yes,
Wikipedia explicitly does not profile you (we have a fairly strong privacy
policy) :)

If people think this is a good idea I might give it a shot.

On Sat, Jun 20, 2015 at 7:15 PM, Jacob Weisz <notifications@github.com>
wrote:

> Well, Wikipedia accounts are, as far as I know, less explicitly tied to
> real identities than Google or Github. And Wikipedia, as far as I know,
> doesn't try to profile you on other sites, which makes it likely for that
> reason a step above Google.
>
> The big one that might be blocking for @kentonv
> <https://github.com/kentonv> is that I've never heard of two-factor
> authentication for Wikipedia accounts.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sandstorm-io/sandstorm/issues/525#issuecomment-113799342>
> .
>



-- 
Yuvi Panda T
http://yuvi.in/blog

**Comment by paulproteus** [comment]
I think it's fascinating, and "obviously" there should be a Sandstorm
instance for the Wikipedia community so that people have a place to run
back-ends for their wiki scripts and more. (I realize someone setting up a
public instance w/ this is a separate idea from the login code, but I
imagine the login code is a pre-requisite.)

I'm curious for Kenton's input, but my feeling is that this is a great idea.

**Comment by kentonv** [comment]
Before we can add a lot of authentication providers, we need to support logging into the same account with multiple providers, that is attaching multiple login mechanisms/identities to one Sandstorm account. Otherwise, people will forget which login provider they had used previously, or just not understand that they aren't equivalent, and then they'll wonder where all their stuff went when they log in. (This is a real problem seen in practice by sites that have lots of login providers.)

Once we've solved that problem, then I'm happy to add all sorts of login providers, provided they have Meteor packages. It looks like there is no accounts-wikipedia page yet, so you could start working on that now, as a Meteor package, even though we won't be able to integrate it just yet. Then when the time comes, it will be trivial for us to turn it on.

**Comment by abliss** [comment]
"attaching multiple login mechanisms/identities to one Sandstorm account" is done now, right? https://sandstorm.io/news/2017-05-08-refactoring-identities

**Comment by zenhack** [comment]
Correct, I think this is unblocked if someone wanted to do the work.

**Comment by kentonv** [comment]
The functionality is there, though experience with Oasis suggests it isn't intuitive. People expect to be able to log in with any provider and get at their stuff, even if they've never gone through the process of adding those credential to their account. It might be nice to add some code that detects if you log in with a new credential whose e-mail address matches a different credential on an existing account, and then asks you if you really meant to use that other credential. If so, it directs you to log in with the original credential, then adds the new one to your account so in the future you can use either.

All that said, yeah, I think my comment from 2015 was before we even supported multiple credential on an account at all (which I think was added in late 2015 -- the 2017 blog post is about a refactoring I did later to make it less confusing).
